worlds/stardew_valley/logic.py

```diff
@@ -1387,10 +1389,93 @@ def can_reach_woods_depth(self, depth: int) -> StardewRule:
         if depth > 30:
             rules.append(self.received("Woods Obelisk"))
         if depth > 50:
-            rules.append(self.has_great_weapon() & self.can_cook())
+            rules.append(self.can_do_combat_at_level("Great") & self.can_cook())
         return And(rules)
 
     def npc_is_in_current_slot(self, name: str) -> bool:
         npc = all_villagers_by_name[name]
         mod = npc.mod_name
         return mod is None or mod in self.options[options.Mods]
+
+# Spell Logic in terms of combat usability (similar to weapons)  Strategy is that if the mod doesn't exist, its
+# always false and thus doesn't get considered relative to other logic tests.
+
+    def can_earn_spells(self) -> StardewRule:
+        return self.has_relationship("Wizard", 3) & self.can_reach_region(SVRegion.pierre_store) & \
+            self.can_reach_region(SVRegion.wizard_tower)
+
+    def has_any_spell(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        return self.can_earn_spells()
+
+    def has_decent_spells(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        magic_resource_rule = self.can_earn_spells() & self.received("Magic Level", 3)
+        magic_attack_options_rule = self.received("Spell: Fireball") | self.received(
+            "Spell: Frostbite") | self.received("Spell: Shockwave") | self.received("Spell: Spirit")
+        magic_support_options_rule = self.received("Spell: Heal") | self.received("Spell: Tendrils")
+        return magic_resource_rule | magic_attack_options_rule | magic_support_options_rule
+
+    def has_good_spells(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        magic_resource_rule = self.can_earn_spells() & self.received("Magic Level", 3)
+        magic_attack_options_rule = self.received("Spell: Fireball") | self.received(
+            "Spell: Frostbite") | self.received("Spell: Shockwave") | self.received("Spell: Spirit")
+        magic_support_options_rule = self.received("Spell: Heal") | self.received("Spell: Tendrils")
+        return magic_resource_rule | magic_attack_options_rule | magic_support_options_rule
+
+    def has_great_spells(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        magic_resource_rule = self.can_earn_spells() & self.received("Magic Level", 6)
+        magic_attack_options_rule = self.received("Spell: Fireball") | self.received(
+            "Spell: Frostbite") | self.received("Spell: Shockwave") | self.received("Spell: Spirit")
+        magic_support_options_rule = self.received("Spell: Heal") | self.received("Spell: Tendrils")
+        return magic_resource_rule & magic_attack_options_rule & magic_support_options_rule
+
+    def has_amazing_spells(self) -> StardewRule:
```

Comment:
Amazing could request more than one offensive spell, maybe all of them or all but one